clicache/src/ExceptionTypes.hpp

```diff
@@ -266,417 +266,3411 @@ namespace Apache
+      };
 
       /// <summary>
       /// Thrown when an argument to a method is illegal.
       /// </summary>
-      _GF_MG_EXCEPTION_DEF3( IllegalArgumentException );
+      [Serializable] public ref class IllegalArgumentException
+          : public GeodeException {
+      public:
+        IllegalArgumentException() : GeodeException() {}
+        IllegalArgumentException(String ^ message) : GeodeException(message) {}
+        IllegalArgumentException(String ^ message,
+                                 System::Exception ^ innerException)
+            : GeodeException(message, innerException) {}
+
+      protected:
+        IllegalArgumentException(SerializationInfo ^ info,
+                                 StreamingContext context)
+            : GeodeException(info, context) {}
+      internal:
+        IllegalArgumentException(
+                       const apache::geode::client::IllegalArgumentException
+                           &nativeEx)
+            : GeodeException(marshal_as<String ^>(nativeEx.getMessage()),
+                             gcnew GeodeException(
+                                 GeodeException::GetStackTrace(nativeEx))) {}
+        IllegalArgumentException(
+            const apache::geode::client::IllegalArgumentException &nativeEx,
+            Exception ^ innerException)
+            : GeodeException(marshal_as<String ^>(nativeEx.getMessage()),
+                             innerException) {}
+        static GeodeException ^
+            Create(const apache::geode::client::Exception &ex,
+                   Exception ^ innerException) {
+              const apache::geode::client::IllegalArgumentException *nativeEx =
+                  dynamic_cast<
+                      const apache::geode::client::IllegalArgumentException *>(
+                      &ex);
+              if (nativeEx != nullptr) {
+                if (innerException == nullptr) {
+                  return gcnew IllegalArgumentException(*nativeEx);
+                } else {
+                  return gcnew IllegalArgumentException(*nativeEx,
+                                                        innerException);
+                }
+              }
+              return nullptr;
+            }
+        virtual std::shared_ptr<apache::geode::client::Exception> GetNative()
+          override {
+          auto message =
+              marshal_as<std::string>(this->Message + ": " + this->StackTrace);
+          if (this->InnerException != nullptr) {
+            auto cause = GeodeException::GetNative(this->InnerException);
+            message += "Caused by: " + cause->getMessage();
+          }
+          return std::make_shared<
+              apache::geode::client::IllegalArgumentException>(message);
+        }
+      };
 
       /// <summary>
       /// Thrown when the state of cache is manipulated to be illegal.
       /// </summary>
-      _GF_MG_EXCEPTION_DEF3( IllegalStateException );
+      [Serializable] public ref class IllegalStateException
```

Comment:
Is there a way to templatize this instead?

Reply:
I guess macros are kind of like templates. Generally old school, but they do help minimize lines of source. This one in particular would be a good one to try and templatize. Will investigate.

Reply:
I don't see a way to templatize the actual name of a class. Compiler gives this error:
     "class template and template parameter cannot have the same name"

This is probably why our native exceptions are also not templated.